server/aesdsocket.c

```diff
@@ -0,0 +1,304 @@
+   {
+       syslog(LOG_ERR, "could not register SIGTERM, error SIG_ERR\n\r");
+       cleanup();
+   }
+
+   bool createdaemon = false;
+    
+   /*check if need to createdaemon*/
+   if (argc == 2) 
+   {
+      if (!strcmp(argv[1], "-d")) 
+      {
+         createdaemon = true;
+      } 
+      else 
+      {
+         printf("wrong arg: %s\nUse -d option for createdaemon", argv[1]);
+         syslog(LOG_ERR, "wrong arg: %s\nUse -d option for createdaemon", argv[1]);
+         return (-1);
+      }
+   }
+
+    struct addrinfo addr_hints;
+
+    memset(&addr_hints, 0, sizeof(addr_hints));
+
+    /*initialise server address*/
+    addr_hints.ai_family = AF_INET;
+    addr_hints.ai_socktype = SOCK_STREAM;
+    addr_hints.ai_flags = AI_PASSIVE;
+    int result = getaddrinfo(NULL, (PORT), &addr_hints, &servaddr);
+    if (result != 0) 
+    {
+       syslog(LOG_ERR, "getaddrinfo() error %s\n", gai_strerror(result));
+       cleanup();
+       return -1;
+    }
+
+    /*create socket connection*/
+    sockfd = socket(servaddr->ai_family, SOCK_STREAM, 0);
+    if (sockfd < 0) 
+    {
+       syslog(LOG_ERR, "socket creation failed, error number %d\n", errno);
+       cleanup();
+       return -1;
+    }
+
+   // Set sockopts for reuse of server socket
+    if (setsockopt(sockfd, SOL_SOCKET, SO_REUSEADDR, &(int){1}, sizeof(int)) < 0) 
+    {
+       syslog(LOG_ERR, "set socket options failed with error number%d\n", errno);
+       cleanup();
+       return -1;
+    }
+
+    // Bind device address to socket
+    if (bind(sockfd, servaddr->ai_addr, servaddr->ai_addrlen) < 0) 
+    {
+       syslog(LOG_ERR, "binding socket error num %d\n", errno);
+       cleanup();
+       return -1;
+    }
+
+    // Listen for connection
+    if (listen(sockfd, MAX_CONNECTIONS)) 
+    {
+       syslog(LOG_ERR, "listening for connection error num %d\n", errno);
+       cleanup();
+       return -1;
+    }
+
+    printf("Listening for connections\n\r");
+
+    if (createdaemon == true) 
+    {
+       int retVal = daemon(0,0);
+       
+       if(-1 == retVal)
+       {
+          syslog(LOG_ERR, "failed to create daemon\n");
+          cleanup();
+          return -1;
+       }
+    }
+
+    while(1) 
+    {
+       struct sockaddr_in cli_addr;
+       socklen_t cli_addr_size = sizeof(cli_addr);
+
+       connfd = accept(sockfd, (struct sockaddr*)&cli_addr, &cli_addr_size);
+        
+       if(connfd < 0)
+       {
+          syslog(LOG_ERR, "accepting new connection error is %s", strerror(errno));
+          cleanup();
+          return -1;
+       } 
+       
+       char *client_ip = inet_ntoa(cli_addr.sin_addr);
+
+       
+       syslog(LOG_INFO, "Accepted connection from %s \n\r",client_ip);
+       printf("Accepted connection from %s\n\r", client_ip);
+
+       char buf[BUF_SIZE];
+
+       /*read the data from client and write into /var/tmp/aesdsocketdata file*/
+       while(1)
+       {
+          int noOfBytesRead = read(connfd, buf, (BUF_SIZE));
```

Comment:
@AnanthD21 what will happen if the packet is more than 1024 bytes before you receive your first '\n'? also does this comply with the requirement as stated in https://www.coursera.org/learn/advanced-embedded-software-development/supplement/N9ZTJ/assignment-5-part-1-instructions which says that the file should be written after newline? This implementation would not work for assignment 8